implement protocol 2.0; change db schema

SomberNight opened the pull request:
This implements Electrum Protocol 2.0, see https://github.com/spesmilo/electrumx/pull/90 (formerly named version 1.5).
Supersedes https://github.com/spesmilo/electrumx/pull/80 (see existing comments there).
Includes https://github.com/spesmilo/electrumx/pull/109.

ahmedbodi committed:
[AuxPow] Add Support for individual block headers instead of a combined hex string

SomberNight committed:
history db: change schema, and rm compaction

SomberNight committed:
history db: change schema: prefix entries with b'H'

SomberNight committed:
history db: store a tx_hash->tx_num map

SomberNight committed:
utxo db: change schema: no longer use compressed txid

SomberNight committed:
db: rm upgrade logic

with the pending db changes, an upgrade is ~as fast as a resync from genesis

SomberNight committed:
history db: sort key insertion order

Supposedly it makes a difference (see e.g. [0]), and depending on how batching
works it makes sense it would, but during a few full syncs of testnet I've done,
it was within measurement error. Still, existing code was already doing this.

[0]: https://stackoverflow.com/q/54941342

SomberNight committed:
session: blockchain.transaction.get_merkle: make "height" optional

now that we have our own txindex

SomberNight committed:
env: add new cfg option "DAEMON_HAS_TXINDEX", and support "False"

SomberNight committed:
history db: store a (prev_txnum, prev_idx)->spender_txnum map

This will allow looking up which tx spent an outpoint.

SomberNight committed:
(trivial) make TXNUM_PADDING a global

SomberNight committed:
db: add parameter for TXOUTIDX_LEN

SomberNight committed:
db: change TXOUTIDX_LEN from 4 to 3, to save db storage size

In Bitcoin consensus, a txout index is stored as a uint32_t.
However, in practice, an output in a tx uses at least 10 bytes (for an OP_TRUE output), so
- to exhaust a 2 byte namespace, a tx would need to have a size of at least 2 ** 16 * 10 = 655 KB,
- to exhaust a 3 byte namespace, a tx would need to have a size of at least 2 ** 24 * 10 = 167 MB.

SomberNight committed:
session: implement "blockchain.outpoint.subscribe" RPC

notifications not implemented yet

SomberNight committed:
"blockchain.outpoint.subscribe" RPC: implement notifications

SomberNight committed:
protocol change: server.version must be the first message on the wire

related: https://github.com/spesmilo/electrumx/pull/8

SomberNight committed:
(bugfix) db: change tx_num endianness (LE->BE) to match db comparator

History.get_txnums and History.backup depend on ordering of tx_nums,
so we want the lexicographical order (used by leveldb comparator) to
match the numerical order.

SomberNight committed:
protocol change: define order of mempool txs in status of scripthash

Note that this is a soft fork: the server can apply it even for past
protocol versions.

Previously, with the order being undefined, if an address had multiple
mempool transactions touching it, switching between different servers
could result in a change in address status simply as a result of these
servers ordering mempool txs differently. This would result in the client
re-requesting the whole history of the address.

-----

D/i | interface.[electrum.blockstream.info:60002] | <-- ('blockchain.scripthash.subscribe', ['660b44502503064f9d5feee48726287c0973e25bc531b4b8a072f57f143d5cd0']) {} (id: 12)
D/i | interface.[electrum.blockstream.info:60002] | --> 9da27f9df91e3f860212f65b736fa20a539ba6e3d509f6370367ee7f10a4d5b0 (id: 12)
D/i | interface.[electrum.blockstream.info:60002] | <-- ('blockchain.scripthash.get_history', ['660b44502503064f9d5feee48726287c0973e25bc531b4b8a072f57f143d5cd0']) {} (id: 13)
D/i | interface.[electrum.blockstream.info:60002] | --> [
{'fee': 200, 'height': 0, 'tx_hash': '3ee6d6e26291ce360127fe039b816470fce6eeea19b5c9d10829a1e4efc2d0c7'},
{'fee': 239, 'height': 0, 'tx_hash': '9e050f09b676b9b0ee26aa02ccee623fae585a85d6a5e24ecedd6f8d6d2d3b1d'},
{'fee': 178, 'height': 0, 'tx_hash': 'fb80adbf8274190418cb3fb0385d82fe9d47a844d9913684fa5fb3d48094b35a'},
{'fee': 200, 'height': 0, 'tx_hash': '713933c50b7c43f606dad5749ea46e3bc6622657e9b13ace9d639697da266e8b'}
] (id: 13)

D/i | interface.[testnet.hsmiths.com:53012] | <-- ('blockchain.scripthash.subscribe', ['660b44502503064f9d5feee48726287c0973e25bc531b4b8a072f57f143d5cd0']) {} (id: 12)
D/i | interface.[testnet.hsmiths.com:53012] | --> f7ef7237d2d62a3280acae05616200b96ad9dd85fd0473c29152a4a41e05686c (id: 12)
D/i | interface.[testnet.hsmiths.com:53012] | <-- ('blockchain.scripthash.get_history', ['660b44502503064f9d5feee48726287c0973e25bc531b4b8a072f57f143d5cd0']) {} (id: 13)
D/i | interface.[testnet.hsmiths.com:53012] | --> [
{'tx_hash': '9e050f09b676b9b0ee26aa02ccee623fae585a85d6a5e24ecedd6f8d6d2d3b1d', 'height': 0, 'fee': 239},
{'tx_hash': 'fb80adbf8274190418cb3fb0385d82fe9d47a844d9913684fa5fb3d48094b35a', 'height': 0, 'fee': 178},
{'tx_hash': '3ee6d6e26291ce360127fe039b816470fce6eeea19b5c9d10829a1e4efc2d0c7', 'height': 0, 'fee': 200},
{'tx_hash': '713933c50b7c43f606dad5749ea46e3bc6622657e9b13ace9d639697da266e8b', 'height': 0, 'fee': 200}
] (id: 13)

SomberNight committed:
"blockchain.outpoint.subscribe" RPC: add optional "spk_hint" argument

SomberNight committed:
session: wait for protocol ver negotiation to finish before other reqs

We already enforce that server.version must be the first received message
in a session, but we also need to ensure that the server finishes processing
that message and sets up the correct protocol version before starting to
process further messages.

SomberNight committed:
new scripthash status: impl protocol 1.5 status; inefficient for now

SomberNight committed:
"blockchain.outpoint.subscribe" RPC: distinguish heights "-1" and "0"

Similar to scripthash statuses, the height of an unconfirmed tx is:
- `-1` if it has any unconfirmed parents,
- `0` otherwise.

SomberNight committed:
DB storage: implement iterator.seek(key)

in a way that works consistently between LevelDB and RocksDB.

SomberNight committed:
new scripthash status: speed-up calc by storing intermed. state in db

SomberNight committed:
"scripthash.get_history": impl updates for proto 1.5; no client_* yet

Handling of client_statushash and client_height is not yet implemented.

SomberNight committed:
"scripthash.get_history": handle client_statushash and client_height

SomberNight commented:
Note: this is ~done; but we probably won't merge it until also implementing the protocol changes in the client, to avoid having to change the protocol again due to some unforeseen issue (though I don't expect issues).
I'll implement the changes in Electrum after the 4.1 release.

JeremyRand committed:
Array headers: fix type error

JeremyRand committed:
Array headers: move variable initialization

JeremyRand committed:
Array headers: Refactor AuxPoW

JeremyRand committed:
Array headers: Refactor AuxPoW truncation

JustinTArthur commented:
New schema looks good. At first attempt at syncing with this, performance isn't great, unfortunately. On AWS' r6g.large instance (dual-core ARM64 w/ SHA extensions, 16GB RAM) with pypy and rocksdb, current spesmilo:master syncs in under 24 hours, whereas this branch is estimated to take about a week. I don't know why CPU isn't able to max out, nor is disk throughput. It might be some very large pages being used in memory, hard to tell.

<img width="1595" alt="electrumx-202010_db2-vs-master-20210602" src="https://user-images.githubusercontent.com/577312/120433182-8a447e80-c340-11eb-940f-961b1659fd2d.png">

Anything else you want me to try, e.g. on the same compute configuration, let me know.

SomberNight commented:
What value of `CACHE_MB` did you test with? If default, it might be worth trying with 2x that.

(The cache stores a lot more data for the history dbs now:
see history.py/`unflushed_memsize` https://github.com/spesmilo/electrumx/pull/101/files#diff-55c36e9a34bb1ea334f60433845bca3c631058fdf83cbe6dc662a531d8a1b35fR211)

aguycalled commented:
hi, what is the expected timeline to see protocol 1.5 merged in electrumx and implemented in electrum? i have a few wallets which can't load because of its very long transaction history

0racl3z reviewed (approved):
Thanks!

SomberNight committed:
session: rework statushash calculation.

If a client requested the status of a very busy address (with cold cache:
no precalc yet), we might have gotten a timeout before we stored any
intermediate status calculated. Then, if the client reconnected and
requested the same, we would get stuck in this loop and never make progress.

With this change, we try to store intermediate hashes sooner, so that even if
there is a timeout, if the client reconnects there is less work to be done.

JeremyRand commented:
> I'll implement the changes in Electrum after the 4.1 release.

@SomberNight Electrum 4.2.1 is out; are there any plans regarding when Protocol 1.5 will be added to Electrum?

JustinTArthur commented:
For client devs wanting to test the new protocol, here are electrumx servers:
**Bitcoin testnet** - db2.electrum.justinarthur.com:51001  
**Bitcoin mainnet** - db2.electrum.justinarthur.com:50001  
Both use TLS. mainnet server is still syncing and may not open its port for 48-ish hours. Use mainnet server with caution… this is new tech being tested.

Metrics and logs for both servers: https://db2.electrum.justinarthur.com/  
use `electrum_dev` as the username and password.

It's not automatically pulling on branch changes so ping me on IRC if any of you need me to pull in new changes, a new branch, or start a new sync. I'm using my cloud vendor's spot market, so servers are subject to occasional interruption if my cloud vendor's customers increase their regularly contracted usage.

JustinTArthur commented:
I'm shutting down the test servers to save money. If anyone is actively developing on Electrum or similar stratum client for protocol 1.5 and needs infrastructure to test against, just ping me on here or as jarthur in #electrum on Libera.

gzk-x64 commented:
Hello @JustinTArthur
Thank you for everything you've done.
Would you mind turning your server back on? I would like to test the v2 protocol with btc-blockchain-explorer to see if the pagination of the numerous transactions works before migrating. 
Otherwise, does anyone have the answer to my question?
